### PR TITLE
f-checkout@v3.24.0 - Address service to correctly map international address lines

### PR DESCRIPTION
**Pull request by FayeCarter**
### Changed
- Address service to correctly map international address lines.

### More Info
UK and NZ have 2 lines of address, city and postcode
AU has the same but with an administrative area
IT has 4 lines of address, city and postcode

We also use of couple of APIs that we call to get the address alongside local storage to retrieve addresses. These are all in different formats. For example:
```
address: {
  lines: [line1, line2]
  postcode: ''
  administrativeArea: ''
}
```
And
```
address: {
  Line1,
  Line2,
  Line3,
  Line4,
  PostalCode,
}
```
Administrative area is also returned in different ways either as `administrativeArea` or as `Line4`

This PR adds tenant specific files to format the addresses we receive to make them usable in Checkout

### CODE REVIEW

File: packages/components/pages/f-checkout/src/services/addressService/index.js [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        };
+    }
+
+    return {
+        ...address,
+        ...addressLines
+    };
+}
+
+function getDefaultAddress (addressData, tenant) {
```
Comment (FayeCarter) [inline packages/components/pages/f-checkout/src/services/addressService/index.js:54]: Not a new, just moved to a different file

File: packages/components/pages/f-checkout/src/services/addressService/index.js [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    }
+
+    return address || addresses?.find(a => a && a.ZipCode.startsWith(postcode));
+}
+
+/**
+ * Checks whether the user has an address in local storage available to us
+ * @returns {boolean} - whether the address is in local storage or now
+ */
+const isAddressInLocalStorage = () => {
```
Comment (FayeCarter) [inline packages/components/pages/f-checkout/src/services/addressService/index.js:93]: Not a new, just moved to a different file

File: packages/components/pages/f-checkout/src/services/addressService/index.js [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+        if (storedAddress) {
+            return mapped ? tenants[tenant].formatAddress(formatAddressLines(storedAddress)) : storedAddress;
+        }
+    }
+
+    return null;
+}
+
+function getAddressCoordsFromLocalStorage () {
```
Comment (FayeCarter) [inline packages/components/pages/f-checkout/src/services/addressService/index.js:125]: Not a new, just moved to a different file

File: packages/components/pages/f-checkout/src/services/addressService/index.js [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+const isAddressInLocalStorage = () => {
+    if (window.localStorage) {
+        const address = window.localStorage.getItem(storedLocationKey) ? JSON.parse(window.localStorage.getItem(storedLocationKey)) : null;
+        return !!(address && (address.PostalCode || address.ZipCode));
+    }
+
+    return false;
+};
+
+function getStoredAddress () {
```
Comment (FayeCarter) [inline packages/components/pages/f-checkout/src/services/addressService/index.js:102]: New

File: packages/components/pages/f-checkout/src/services/addressService/index.js [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    return storedLocation ? JSON.parse(storedLocation) : null;
+}
+
+/**
+ * Retrieves the address from local storage
+ * @param mapped - boolean deciding whether to return the address in a mapped state
+ * @returns address - The address from local storage
+ */
+function getAddressFromLocalStorage (tenant, mapped = true) {
+    if (window.localStorage) {
```
Comment (FayeCarter) [inline packages/components/pages/f-checkout/src/services/addressService/index.js:114]: Not a new, just moved to a different file

File: packages/components/pages/f-checkout/src/services/addressService/index.js [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+}
+
+/**
+ * Checks whether the address values in local storage match what's is in the form.
+ * @param storedAddress - The address stored in local storage
+ * @param formAddress - The address in form state
+ * @returns {boolean} - Whether the form values match
+ */
+function doesAddressInStorageAndFormMatch (storedAddress, formAddress) {
+    return storedAddress.Line1 === formAddress.line1 && storedAddress.PostalCode === formAddress.postcode
```
Comment (FayeCarter) [inline packages/components/pages/f-checkout/src/services/addressService/index.js:142]: Not a new, just moved to a different file

File: packages/components/pages/f-checkout/src/services/addressService/tenants/uk.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
+function isFullPostCode (postcode) {
+    return !postcode ? false : postcode.length >= 5;
+}
+
+function toFormattedPostcode (postcode) {
+    let formattedPostcode = postcode.toString().replace(/\s/g, '').replace(/-/g, '');
```
Comment (FayeCarter) [inline packages/components/pages/f-checkout/src/services/addressService/tenants/uk.js:6]: Not a new, just moved to a different file

File: packages/components/pages/f-checkout/src/services/addressService/index.js [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+import it from './tenants/it';
+
+const tenants = {
+    uk,
+    au,
+    nz,
+    it
+};
+
+const storedLocationKey = 'je-full-address-details';
```
Comment (FayeCarter) [inline packages/components/pages/f-checkout/src/services/addressService/index.js:13]: Not a new, just moved to a different file

File: packages/components/pages/f-checkout/src/services/addressService/index.js [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+
+    if (!postcode) {
+        return '';
+    }
+
+    return tenants[tenant].toFormattedPostcode(postcode);
+}
+
+function formatAddressLines (address) {
+    let addressLines;
```
Comment (FayeCarter) [inline packages/components/pages/f-checkout/src/services/addressService/index.js:30]: new

File: packages/components/pages/f-checkout/src/services/addressService/index.js [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+        }
+
+        const defaultAddressId = addressData.DefaultAddress;
+
+        return address.AddressId === defaultAddressId;
+    });
+}
+
+function getAddressClosestToPostcode (currentPostcode, addressData, tenant) {
+    const addresses = addressData.Addresses;
```
Comment (FayeCarter) [inline packages/components/pages/f-checkout/src/services/addressService/index.js:69]: Not a new, just moved to a different file

File: packages/components/pages/f-checkout/src/services/addressService/tenants/au.js [hunk 11]
```diff
@@ -0,0 +1,29 @@
+function isFullPostCode (postcode) {
+    return !postcode ? false : postcode.length === 4;
+}
+
```
Comment (FayeCarter) [inline packages/components/pages/f-checkout/src/services/addressService/tenants/au.js:4]: Checkout handles 2 addresses from two apis and local storage. All results are formatted slightly differently. Each tenant requires slightly different address lines.

Tenant files have been added to format the results from each of these sources to the lines required per country

File: packages/components/pages/f-checkout/src/services/addressService/index.js [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+const storedLocationKey = 'je-full-address-details';
+
+function toFormattedPostcode (postcode, tenant) {
+    const shouldFormatPostCode = tenant === 'uk';
+
+    if (!shouldFormatPostCode) {
+        return postcode;
+    }
+
+    if (!postcode) {
```
Comment (matthewhardern) [inline packages/components/pages/f-checkout/src/services/addressService/index.js:18]: this being after the above could mean that you return an `undefined` postcode if shouldFormatPostCode is false, is this intended?

Comment (FayeCarter) [reply packages/components/pages/f-checkout/src/services/addressService/index.js:18]: It wasn't intentional, thanks!

**Comment by matthewhardern** [comment]
Based on recent call it LGTM, just have to see if there any pain points in future, happy to approve for now.

**Review by matthewhardern — APPROVED** [review]

**Review by BanksySan — APPROVED** [review]

File: packages/components/pages/f-checkout/src/store/checkout.module.js [hunk 13]
```diff
@@ -498,17 +506,14 @@ export default {
...
                 /* eslint-disable prefer-destructuring */
                 address = fulfilment.location.address;
             }
 
             if (address) {
-                /* eslint-disable prefer-destructuring */
-                state.address.line1 = address.lines[0];
-                state.address.line2 = address.lines[1];
-                /* eslint-enable prefer-destructuring */
+                [state.address.line1, state.address.line2, state.address.line3, state.address.line4] = address.lines;
```
Comment (charliearlie) [inline packages/components/pages/f-checkout/src/store/checkout.module.js:516]: Comment to link to PI-14068

**Pull request merged**